### ISSUE TITLE
Proposal: lightweight hidden-state probing API for interpretability

### ISSUE THREAD

**Issue by julienlafrance** [soda-inria/tabicl#102]
Hi,

I've been working with TabICL for an interpretability project — Neural Collapse / ETF emergence in in-context attention, representation similarity across upstream encoders. Probing internals currently requires reaching into `clf.model_.icl_predictor` manually and working around sklearn wrapper side effects (label encoding, ensemble class permutations, hidden-state re-ordering).

Related: issue #97 asks for checkpoint access without the classifier wrapper. This proposal goes one step further — a small utility to extract internal hidden states while keeping the wrapper functional.

Context: I've contributed GPU-parallelization work to TabPFN before (https://github.com/PriorLabs/tabpfn-extensions/pull/257), and come to TabICL with the same mindset — minimal overhead, GPU-native, clean API.

Would the team welcome a small contribution adding a probing API? Rough sketch:

```python
from tabicl.probe import HiddenExtractor

ext = HiddenExtractor.from_pretrained(device="cuda:0")

# single forward, rows aligned to input
H = ext.extract_tf_icl(X_train, y_train, X_test)     # (T, d_model)

# arbitrary intermediate block
H_b6 = ext.extract(X_train, y_train, X_test, layer="blocks.6")

# per-estimator extraction for ensemble analysis
H_ens = ext.extract_per_estimator(X, y, X_test, n_estimators=8)  # (E, T, D)
```

**Use cases**:
- Neural Collapse / ETF measurement on hidden states
- representation similarity across upstream encoders (Platonic-style)
- linear-probing intermediate features
- logit lens / attention-pattern analysis

**Closest prior art**: Pinto 2025 (arXiv:2506.06443) and Ben Hicham et al. 2026 (arXiv:2604.16123) both use tabular FMs as black-box classifiers; no published tool specifically for probing TabICL internals.

**Design principles**:
- torch-native, bypasses sklearn for minimal overhead (bf16 autocast)
- hook-based on `icl_predictor.tf_icl` and any transformer block
- handles label encoding + ensemble permutations transparently
- numpy-in / numpy-out by default, torch opt-in
- zero new dependencies

Working prototype on my side; happy to share if helpful.

Three questions before polishing a PR:
- (a) Preferred integration — new `tabicl/probe/` submodule, extension of existing sklearn wrapper, or top-level `tabicl.probe` namespace?
- (b) Any existing internal probing utilities to coordinate with?
- (c) Code/test/docs conventions — pytest layout, docstring format, type-hint strictness?

Thanks for TabICL — excellent library.

**Comment by JingangQu** [comment]
Hi,

Of course, welcome contribution!!!  (a) `tabicl/probe` submodule is good. (b) nothing existing to coordinate with at the moment for probing. (c) we largely follow scikit-learn convention.
